### ISSUE TITLE
Switch away from react-bootstrap to a headless component library (Revisit behavior of React-based form validation)

### ISSUE THREAD

**Issue by nwalters512** [PrairieLearn/PrairieLearn#13581]
Here and elsewhere, the input help text is hidden when an error is shown:

https://github.com/PrairieLearn/PrairieLearn/blob/af247e27dcacee5e77a4c0efda3692fe7227b4c1/apps/prairielearn/src/pages/instructorInstanceAdminSettings/components/CopyCourseInstanceModal.tsx#L131-L140

I don't understand why we do this - why hide the helpful explanatory text when the input is empty?

<img width="213" height="102" alt="Image" src="https://github.com/user-attachments/assets/4c6c5ddd-f6f4-4728-8582-3d1cfd3264f3" />

While we're here, I have other questions/concerns:

- I don't understand the choice to use `text-danger small mt-1` instead of `form-text text-danger`. Or, for that matter, why we don't use `invalid-feedback` as in https://github.com/PrairieLearn/PrairieLearn/pull/13504.
- When we use native HTML input validation, we get accessibility for free; not so with `react-hook-form` and such. At a minimum, I think we need to be adding `aria-invalid` attributes and programmatically associating error messages with the input.
  - Aside: this is one of the reasons I'm skeptical of using `react-hook-form` over native browser validation. Slapping a `pattern` or `required` onto an input makes it trivial to fall into the pit of success. Using other libraries requires a lot of careful implementation, and we're failing to do so correctly.

**Comment by nwalters512** [comment]
@mwest1066 and I discussed the error vs. description text a bit today and lean towards always showing both (possibly with some tuning to avoid having the error messages be redundant - though this may not be necessary). We'd rather reliably show more (enough) information than either not show enough, or do a lot of careful thinking to avoid showing any duplicate info.

**Comment by reteps** [comment]
A few thoughts:

1. I agree that we should make a way to have a input with a nice UI, and good accessibility. I think we should make a component for this. `react-hook-form` is easy to integrate here if we want (https://react-aria.adobe.com/forms#react-hook-form).

2. I think we should aim for a solution that looks nice and is accessible. Moving forward into components seems better than stepping back to native form validation. If you are unconvinced, I can do a survey of the two UI patterns in Slack among instructors and see which they prefer. Having a red error message appear below the input is (IMO) expected UX.

2. What do we base the component (e.g. a `FormInput` in `@prairielearn/ui`) on? We are reaching the point where the bootstrap styles are much less complex than proper aria handling. There are other headless libraries that have extensive aria support. Mainly the two current frontrunners are: https://base-ui.com/ and https://react-aria.adobe.com/. I would prefer to base the component on `FormInput`

3. Originally, I used `react-hook-form` as the verbosity of `tanstack form` was ridiculous. I recently saw that Tanstack supports a "component library" style pattern. See https://x.com/ugbahisioma/status/1999937172955369824?s=46. We could consider using tanstack form.

**Comment by reteps** [comment]
In the meeting today, we discussed that we want to explore a library that has built-in ARIA handling, and that I should prototype two components (one with each library) so we can get a feel for the differences. We want to stick with react-hook-form.

**Comment by reteps** [comment]
After some exploration today:

1. React-aria has many more components. These are planned in base-ui, but not here yet. https://github.com/orgs/mui/projects/1
2. React-aria has some opinionated hooks for accessibility they recommend using (https://react-aria.adobe.com/usePress). Let's discuss this core idea they have -- which is using hooks for state.
3. React-aria is still getting major updates every month, even though it's a large project.  https://react-aria.adobe.com/releases/v1-14-0

Overall, I think React-aria is a better bet for long term maintenance.

**Comment by reteps** [comment]
@nwalters512 and I discussed that we like `react-aria`. This will happen as a gradual, as-needed migration. I am closing this out as this doesn't track any actionable items for me anymore.
